alternator/executor.cc

```diff
@@ -1636,19 +1649,56 @@ future<executor::request_return_type> rmw_operation::execute(service::storage_pr
     });
 }
 
+bytes query_cache_key(const std::string_view& query, char query_type) {
+   auto b = md5_hasher::calculate(query);
+   b[0] = query_type; // this avoids key collision for different expression types
+   return b;
+}
+
+template<typename T, typename F>
+requires requires(F f, std::string_view v) {
+    { f(v) } -> std::same_as<T>;
+}
+static T parse_or_get_from_cache(lw_shared_ptr<expression_cache_t> cache, F parse_func, std::string_view query, char query_type) {
+    auto key = query_cache_key(query, query_type);
+    auto ptr = cache->find(key);
+    if (ptr) {
+        // we need to copy, it can be evicted from cache anytime and
+        // aditonally our expression evaluation code modifies T
```

Comment:
What's T here?

Why does the expression evaluation code modify T (I presume some expression type)?

Reply:
I don't know what is "T" (I can't see now the context), but I think what Marcin was trying to say that in Alternator expressions, we "resolve" the expression (similar to CQL's "binding" or "preparing")  by modifying the expression in-place, so the caller needs to get a copy of the expression that they are allowed to modify - not a pointer into the cache where we don't want to ruin the parsed but yet-unresolved expression.

Since expressions are typically not huge, I think that copying them is a small price to pay.